opentelemetry-sdk/src/opentelemetry/sdk/_metrics/aggregation.py

```diff
@@ -126,12 +131,7 @@ def aggregate(self, value):
         if isinstance(self._instrument, _Monotonic):
             self._sum += value
 
-        for key in self._value.keys():
-
-            if value < key:
-                self._value[key] = self._value[key] + value
-
-                break
+        self._value[self._boundaries[bisect(self._boundaries, value)]] += 1
```

Comment:
Is there an off by one error here and we should be using `bisect_left()` instead? From the [proto](https://github.com/open-telemetry/opentelemetry-proto/blob/8672494217bfc858e2a82a4e8c623d4a5530473a/opentelemetry/proto/metrics/v1/metrics.proto#L554-L567):

```proto
  // explicit_bounds specifies buckets with explicitly defined bounds for values.
  //
  // This defines size(explicit_bounds) + 1 (= N) buckets. The boundaries for
  // bucket at index i are:
  //
  // (-infinity, explicit_bounds[i]] for i == 0
  // (explicit_bounds[i-1], explicit_bounds[i]] for 0 < i < N-1
  // (explicit_bounds[i], +infinity) for i == N-1
  //
  // The values in the explicit_bounds array must be strictly increasing.
  //
  // Histogram buckets are inclusive of their upper boundary, except the last
  // bucket where the boundary is at infinity. This format is intentionally
  // compatible with the OpenMetrics histogram definition.
```

Just playing around with this, a hist with boundaries `[0, 1, 2]` should have buckets:

0: (-inf, 0]
1: (0, 1]
2: (1, 2]
3: (2, inf)

but bisect is off by one at the boundaries:
```py
>>> boundaries = [0, 1, 2]
>>> bisect.bisect(boundaries, 0)
1
>>> bisect.bisect(boundaries, 1)
2
>>> bisect.bisect(boundaries, 2)
3
```

I could be misreading the code, is there a test case covering this?

Reply:
Ah, I see. This is a great observation. I now think that it is better to use a list instead of a dictionary to keep track of the counters. It is more similar to the proto and also, the dictionary is ambiguous, for example: `counters[4] == 9`. Does that mean that the bucket left or right of the boundary `4` has a value of `9`?

I prefer to fix this issue in this PR as well since it now makes little sense to keep the `OrderedDict`.

Reply:
Fixed and added a test case